ONSdigital/eq-author-app
# EAR 903 - Update front-end GraphQL for Folders

## What is the context of this PR?

Fixes front-end graphql so that Author works as-is with the back-end folders changes. 
Provides a base from which front-end folder functionality can be implemented on top of.

### 🌵 The Good
##### Questionnaires
- Viewing list of questionnaires
- Creating a questionnaire
- Opening a questionnaire
- Deleting a questionnaire
- Duplicating a questionnaire
- Viewing QCodes page
- Viewing Sharing / Settings pages
- Validation error subscription

##### Pages
- Viewing a question page
- Creating a question page using the button on the question design page
- Creating a question page using the nav bar menu (context dependent as before)
- Calculated summary pages
- Duplicating a question page
- Deleting a question page
- Creating a calculated summary page
- Deleting a calculated summary page
- Creating a confirmation page
- Deleting a confirmation page
- Moving pages in same section
- Moving pages between sections - including generating replacement page when empty (now handled by server rather than client)

##### Answers
- Creating all answer types
- Updating answers
- Updating validation for answers

##### Sections
- Creating a section
- Duplicating a section
- Deleting a section, including deleting the final section

##### Logic
- The "No Routing" page
- Adding routing logic
- Adding skip logic
- Choosing routing destinations

### 🌵 The Bad
... Hopefully dealt with for now 🔫

### 🌵 The Ugly
- The move page functionality has a quick fix for now - flattening all folders into one list for the purpose of choosing positions. Will need re-doing properly when we get to the appropriate ticket.
- `optimisticResponse` removed for now from the `withMovePage` HOC - may be re-added later (if required) once that functionality is appropriately adjusted to be target folder aware.

## Back-end changes
- When a folder has all its questions deleted, delete the folder - unless it's the only folder left in the section. This was necessary as it was otherwise impossible to delete an only child question (of a folder) from the questionnaire.
- Resolvers for `QuestionPage.position` and `CalculatedSummaryPage.position` updated to return a page's position within parent folder rather than relative to the section.
- `duplicatePage` resolver updated so that if a question in a disabled folder is duplicated, the copied question is added to a new disabled folder. This is to keep consistency with other behaviour - newly coined questions are created in empty disabled folders by default. This change allows move page logic to work as is on the client and hopefully will simplify the implementation of the create folder UI etc.
- Added ability to query a page's parent `folder`.
- Added ability to query a folder's parent `section`.

## Publisher v2
- Publisher v2 has been modified just enough to work with folders in situ. Folders are mapped to groups in the Runner schema. Disabled folders are merged together to save producing needlessly large amounts of groups.
- Will need future work during the "add skip logic to folders" ticket.
- Matching changes made in PR for v3: https://github.com/ONSdigital/eq-publisher-v3/pull/15

## How to review

- Load up Author, create a questionnaire
- Try everything and see if it breaks
- Check that you can view the questionnaire in Runner

## What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process
